framework/util/page_guard_manager.cpp

```diff
@@ -205,33 +205,83 @@ PageGuardManager::PageGuardManager() :
     enable_separate_read_(kDefaultEnableSeparateRead), unblock_sigsegv_(kDefaultUnblockSIGSEGV),
     enable_signal_handler_watcher_(kDefaultEnableSignalHandlerWatcher),
     signal_handler_watcher_max_restores_(kDefaultSignalHandlerWatcherMaxRestores),
-    enable_read_write_same_page_(kDefaultEnableReadWriteSamePage)
+    enable_read_write_same_page_(kDefaultEnableReadWriteSamePage), protection_mode_(kDefaultMemoryProtMode)
 {
-    InitializeSystemExceptionContext();
+    if (kUserFaultFdMode == protection_mode_ && !USERFAULTFD_SUPPORTED)
+    {
+        GFXRECON_LOG_WARNING(
+            "Kernel does not support all required userfaultfd features. Falling back to mprotect method");
+
+        protection_mode_ = kMProtectMode;
+    }
+
+    if (kMProtectMode == protection_mode_)
+    {
+        InitializeSystemExceptionContext();
+    }
+    else
+    {
+        if (!InitializeUserFaultFd())
+        {
+            GFXRECON_LOG_ERROR("Userfaultfd initialization failed. Falling back to mprotect method");
+
+            protection_mode_ = kMProtectMode;
+            InitializeSystemExceptionContext();
+        }
+    }
 }
 
-PageGuardManager::PageGuardManager(bool enable_copy_on_map,
-                                   bool enable_separate_read,
-                                   bool expect_read_write_same_page,
-                                   bool unblock_SIGSEGV,
-                                   bool enable_signal_handler_watcher,
-                                   int  signal_handler_watcher_max_restores) :
+PageGuardManager::PageGuardManager(bool                 enable_copy_on_map,
+                                   bool                 enable_separate_read,
+                                   bool                 expect_read_write_same_page,
+                                   bool                 unblock_SIGSEGV,
+                                   bool                 enable_signal_handler_watcher,
+                                   int                  signal_handler_watcher_max_restores,
+                                   MemoryProtectionMode protection_mode) :
     exception_handler_(nullptr),
     exception_handler_count_(0), system_page_size_(util::platform::GetSystemPageSize()),
     system_page_pot_shift_(GetSystemPagePotShift()), enable_copy_on_map_(enable_copy_on_map),
     enable_separate_read_(enable_separate_read), unblock_sigsegv_(unblock_SIGSEGV),
     enable_signal_handler_watcher_(enable_signal_handler_watcher),
     signal_handler_watcher_max_restores_(signal_handler_watcher_max_restores),
-    enable_read_write_same_page_(expect_read_write_same_page)
+    enable_read_write_same_page_(expect_read_write_same_page), protection_mode_(protection_mode)
 {
```

Comment:
This constructor and the previous constructor used to be one line.  I think these two may now be candidates to share a function which does the shared work.  Or could one use the other as a delegated constructor?  I think delegated constructors came in with C++11.